source/extensions/stat_sinks/hystrix/hystrix.cc

```diff
@@ -0,0 +1,375 @@
+#include "extensions/stat_sinks/hystrix/hystrix.h"
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <sstream>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+#include "common/http/headers.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace StatSinks {
+namespace Hystrix {
+
+const uint64_t HystrixSink::DEFAULT_NUM_OF_BUCKETS;
+
+ClusterStatsCache::ClusterStatsCache(const std::string& cluster_name) {
+  const std::string cluster_name_with_prefix = absl::StrCat("cluster.", cluster_name, ".");
+
+  upstream_rq_2xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_2xx");
+  upstream_rq_4xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_4xx");
+  retry_upstream_rq_4xx_name_ = absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_4xx");
+  upstream_rq_5xx_name_ = absl::StrCat(cluster_name_with_prefix, "upstream_rq_5xx");
+  retry_upstream_rq_5xx_name_ = absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_5xx");
+
+  errors_name_ = absl::StrCat(cluster_name_with_prefix, "errors");
+  success_name_ = absl::StrCat(cluster_name_with_prefix, "success");
+  total_name_ = absl::StrCat(cluster_name_with_prefix, "total");
+  timeouts_name_ = absl::StrCat(cluster_name_with_prefix, "timeouts");
+  rejected_name_ = absl::StrCat(cluster_name_with_prefix, "rejected");
+}
+
+// Add new value to rolling window, in place of oldest one.
+void HystrixSink::pushNewValue(RollingWindow& rolling_window, uint64_t value) {
+  if (rolling_window.empty()) {
+    rolling_window.resize(window_size_, value);
+  } else {
+    rolling_window[current_index_] = value;
+  }
+}
+
+uint64_t HystrixSink::getRollingValue(RollingWindow rolling_window) {
+
+  if (rolling_window.empty()) {
+    return 0;
+  }
+  // If the counter was reset, the result is negative
+  // better return 0, will be back to normal once one rolling window passes.
+  if (rolling_window[current_index_] < rolling_window[(current_index_ + 1) % window_size_]) {
+    return 0;
+  } else {
+    return rolling_window[current_index_] - rolling_window[(current_index_ + 1) % window_size_];
+  }
+}
+
+void HystrixSink::updateRollingWindowMap(Upstream::ClusterInfoConstSharedPtr cluster_info,
+                                         Stats::Store& stats) {
+  const std::string cluster_name = cluster_info->name();
+  Upstream::ClusterStats& cluster_stats = cluster_info->stats();
+
+  if (cluster_stats_cache_map_.find(cluster_name) == cluster_stats_cache_map_.end()) {
+    // ClusterStatsCache cluster_stats_cache_inst(cluster_name);
+    // cluster_stats_cache_map_[cluster_name] = cluster_stats_cache_inst;
+    cluster_stats_cache_map_[cluster_name] = std::make_unique<ClusterStatsCache>(cluster_name);
```

Comment:
I'm concerned that a per-cluster cache where elements are only added but never deleted is going to effectively be a memory leak.  Also, can you describe what you want to happen if a cluster is deleted and then another cluster is, many flushes later, added under the same name?  IIUC, with the current implementation, this will result in that cluster using the old, unrelated cluster's cache.  I have a few ideas of how to flush the cache, but before proposing anything, I'd like to understand what the desired behavior is for naming collisions is.

Reply:
Good catch, I did not consider those cases. I think the best would be to detect when a cluster no longer exists and remove from the cache map. What are your ideas?

Reply:
I haven't thought it through completely, but I think one solution would be to basically check the differences in sizes in the two maps (the cluster map in the server and the sink's cache map) after you finish updating the caches.  Since you've already added any missing clusters to the sink's map, it's only possible that the sink map is larger.  If it is larger, you can iterate through the sink's map and look up each name in the server's cluster map and delete any that aren't found in the server's map.  It's not an incredibly elegant solution, but I'm not sure if there are any better options for trying to sync two unordered maps.

Reply:
Almost done handling the clusters removal + updating the test. How can I simulate removing and adding back clusters during operation? (not in a test)